src/ai/default/contexts.hpp

```diff
@@ -47,6 +47,7 @@ struct target {
 
 class attack_analysis : public wfl::action_callable
 {
+	attack_analysis* clone() const override {return new attack_analysis(*this);}
```

Comment:
Please use unique_ptr or shared_ptr instead of raw pointers. I had to do a similar conversion in b63b8a959a7b217228d44029466486e135cfdfb5 in order to start refactoring the editor code.

That would mean that the return type couldn't be co-variant, so all these clone() methods would have to return `unique_ptr<wfl::action_callable>` instead of `unique_ptr<attack_analysis>`, however does the caller ever need to know the exact subclass?

Reply:
Are you sure `unique_ptr` is correct though? This pointer is only going to be wrapped in a `shared_ptr`. You _are_ correct that there's no need for covariance here, but given that this is a _private_ clone function and the pointer is immediately wrapped in `shared_ptr`, it kinda feels wrong to use a `unique_ptr`.

Reply:
I did say "unique_ptr or shared_ptr"

Reply:
The result of this function _is_ wrapped in a `shared_ptr` already, and the function is private so it shouldn't be possible to call it anywhere else.

Reply:
The smart pointers are tools to help with tracking ownership, using std::make_... instead of `new` makes the code more readable. Particularly in a PR that's fixing an object lifecycle bug, and where there's still a unit test failure to debug, you should be aiming for better code readability.

Returning shared_ptr and using std::make_shared is more efficient than allocating an object with `new` and then creating a shared_ptr to wrap it. You don't like the idea of returning a unique_ptr and then wrapping it, but using unique_ptr should be the same overhead as returning a raw pointer and then wrapping it - both of them are more expensive than using std::make_shared.

I recommend watching one of Herb Stutter's lectures about smart pointers.